### DISCUSSION TITLE
I can't create my own JK flip flop and counter

### DISCUSSION THREAD

**Discussion by VerdichteteEnergie** [hneemann/Digital#1108] [Q&A] [answered]
My goal is to build my own chips from logic gates and then use them for other circuits.

Unfortunately I can't build a JK flip flop and a counter.
After trying several versions of circuits I didn't get any success.

Now I want start with the simplest options for a JK flip flop and a counter.

Can anyone help me or link to a schematic wich works for digital? I'm obviously too stupid to find anything suitable on google...

Please don't be surprised, I'm a beginner.

**My JK Flip Flop**
![jk-flip-flop](https://user-images.githubusercontent.com/76449084/215146882-cb3dc104-7647-40ad-8c6a-43484a82f149.PNG)

**My Counter**
![counter](https://user-images.githubusercontent.com/76449084/215146898-4cc0094d-20a5-4cd1-9f80-49c8f6b93139.PNG)

**Comment by hneemann** [comment, marked as answer]
The idea for your JK flip-flop is not so bad. You just overlooked that your circuit essentially corresponds to an inverter which is fed back to itself:

![inverter](https://user-images.githubusercontent.com/7190563/215257281-a9705409-c507-4d13-9c19-9dee4101a65c.png)

Such a circuit cannot achieve a stabilized state.

Therefore, this direct feedback must be prevented for the circuit to work.
And there are two ways to do that: One can either make the switching pulse so short that no oscillation can occur, because there is not enough time for that. To do this, the rising edge is transformed into a very short pulse. Just long enough to toggle the flipflop, but not long enough for multiple toggles.

![JK-T](https://user-images.githubusercontent.com/7190563/215257578-4c41d24c-2c96-4ece-a271-c07c71f86957.png)

The three gates directly behind the clock input are used to generate the pulse.

The second possibility is to use both edges, the rising and the falling edge of the clock to forward the signal. This approach leads to a master-slave flip-flop. Since the clock cannot be high and low at the same time, the flipflop is never transparent with respect to changes of an input.

![JK-MS](https://user-images.githubusercontent.com/7190563/215257683-ee0d5e25-416b-453a-ad4b-150b1df940e2.png)

Both implementations can be found in the examples folder (_examples/sequential/JK-T.dig_ and _JK-MS.dig_).

**Reply by VerdichteteEnergie** [reply]
Thank You very much! From now on I always look at the examples.

**Reply by hneemann** [reply]
Don't be afraid to keep asking questions here! :smile:

**Comment by jongena** [comment]
Thank you for the interesting discussion. I tried to study the detailed behaviour of the pulses in the two cases given by Mr. Neeman and then to compare with the non working J-K flip flop shown at the very beginning of the discussion. To see all details I used the 
"Show measurement graph at the beginning of the simulation in single gate mode" [Clicking "Edit" at the top of the Digital screen, then select in the pop-up window "Circuit specific settings"; then in the new window click "Advanced" and in the new pop-up window click "Show measurement graph at the beginning of the simulation in single gate mode"). This then keeps the results of the graph on the screen when the simulation stops. However, this went fine as long as there were no oscillations. See the graphs attached: the first one obtained when stopping the simulation without oscillations (e.g. in this case J high or low, with K low); the second one when oscillations occurred (when putting both J and K inputs high): all past timing info for the inputs J, K, and clock are erased. The outputs Q and not Q also show suddenly something else than that what was on the screen before the oscillations were present.

Q1: What did I overlook ? 
Q2: How to get the right graph (and keep the detailed time history) at the moment the oscillations appear?
Q3: If the answer on Q1 and Q2 is negative now, can this be corrected in a next version of Digital ? 

<img width="823" alt="Single gate graph when stopped without oscillations" src="https://user-images.githubusercontent.com/108168531/230718557-ab1cab2e-0c6b-40ff-a9e1-b647c283e8d0.png">

<img width="809" alt="Single gate graph when stopped with oscillations" src="https://user-images.githubusercontent.com/108168531/230718555-89485fbc-bdfc-4faa-96a2-a1953d92ba3b.png">

**Reply by hneemann** [reply]
Perhaps this helps to understand why the top most circuit does not stablize:

https://user-images.githubusercontent.com/7190563/230720430-bd8ac240-361f-4c4f-bd72-b3951d18f734.mp4

**Reply by jongena** [reply]
Dear Mr. Neemann, many thanks for the very nice illustration, also again on the use of Digital to analyze in fine detail the sequence of pulses in logic circuits. This indeed clarifies. The clock has to go low before a next Q/non-Q toggle occurs otherwise further Q/non-Q toggles will occur and thus the circuit cannot arrive at a stable state. Excellent example !
